\documentclass{article}
\usepackage{fullpage}
\usepackage{color}
\usepackage{amsmath,amssymb,amsthm}
\usepackage{natbib}

\bibliographystyle{plainnat}

\include{macros}

\newcommand{\branch}{\mathcal{B}} % branch length stat
\newcommand{\site}{\mathcal{S}} % site stat

\newcommand{\plr}[1]{{\color{blue} \it #1}}


\begin{document}

\begin{center}
    A general class of tree-based statistics,
    and how to compute them
\end{center}

\emph{Other title ideas:}

Efficient computation of a general class of DNA sequence statistics

Theory and computation of single-site statistics of DNA sequences


%%%%%%%%%%
\paragraph{Abstract}
Here we describe how to efficiently compute single-site statistics
from population genetic data
using efficient methods on tree sequences as implemented in msprime.
The method is general enough to compute any currently-defined statistic,
and we use this general framework to suggest a number of new statistics.
\plr{say what method is: sthg about weighting fns}
This formulation also makes it easy to see what statistic of the underlying genealogical trees
each sequence-based statistic is estimating,
and, if underlying trees are also available,
the same methods can be used to quickly compute these quantities.


%%%%%%%%%%%%%%%%%%%%%%%
\section*{Introduction}

\plr{utility of looking at lots of individual-based stats}
With whole-genome sequence
we can confidently infer quantities
between \emph{individuals} that previously had to be estimated
averaging over populations.
This gives us much higher resolution, both in space and along the genome.
With geographic sampling,
there is potentially a lot of information available
from looking at all pairwise comparisons.

\plr{computational challenges}
Single statistics are not too bad to compute
since you can get them from allele frequencies,
but larger summaries, like the matrix of pairwise divergences between $N$ individuals,
naively involve $O(N^2)$ comparisons.

\plr{Something about how} in many applications
we'd really like to know the trees
and sequence differences really just stand in for these.
Example: $F_{ST}$ is X; 
$\theta_W$ is X;
divergence is X.

\plr{Review of phylogenetics lit}
a good review of theoretical properties \citet{semple_and_steel}

%%%%%%%%%%%%%%%%%%%%%%%
\section*{Methods}

We consider only so-called \emph{single site} statistics.
These are statistics computed as a simple \emph{average} 
of a quantity calcuated across a number of sites.
For a concrete example, to compute an element of the site frequency spectrum --
the proportion of sites at which an allele is present in 10\% of the samples, for instance --
we could 
(a) assign each site a value, either $+1$ if it has an allele at 10\%, or zero otherwise,
and
(b) take the average of these values across all sites.
As another example,
to compute the pairwise divergence between two sequences,
we would similarly assign each site either $+1$ or 0 according to whether the sequences differ,
and average these values.


%%%%%%%%%%%%%%%%%%
\subsection*{Tree sequences}

Quick review of tree sequence terminology
and storage:
nodes,
edges,
sites,
mutations.


%%%%%%%%%%%%%%%%%%
\subsection*{Efficient computation}

First we describe how to define
and efficiently compute a quite general set of functions of a collection of sequences
using tools from tree sequences.
In the next section we will specialize these to a natural set of statistics.

\plr{Since the ``weighting function' can take negtive values, maybe ``weight'' is not a good term.  Instead?}

\paragraph{Weighted site averages}
Suppose that we have $k$ groups of samples, $A_1$, \ldots, $A_k$.
% with $n_1$, \ldots, $n_k$ individuals in each, respectively.
The \emph{weight} we compute for a given site is a sum of weights computed for each allele at that site,
and the weights are found as a function of the vector of numbers of individuals in each group inheriting that allele.
Concretely, we specify a \emph{weighting function} $w(x_1, \ldots, x_k)$,
and we say that the \emph{average site weight} of a set of $L$ sites is
\begin{align} \label{eqn:average_site_weight}
    \site_w := \frac{1}{L} \sum_{i=1}^L \sum_{s \in I_i} w(x_1(s), \ldots, x_k(s)) ,
\end{align}
where $I_i$ is the set of alleles seen at that site and
$x_j(s)$ is the number of individuals in $A_j$ that carry allele $s$.

Before we describe how to compute this,
we give the analogous topological quantity,
that is equal to the expected weighted site average
under a continuum-sites model.

\paragraph{Weighted branch length averages}
The \emph{average branch length weight}
is computed almost just as above:
\begin{align} \label{eqn:average_branch_weight}
    \branch_w :=  \frac{1}{L} \sum_{i=1}^L \sum_{e \in T_i} \ell(e) w(x_1(e), \ldots, x_k(e)) ,
\end{align}
except that now $T_i$ is the genealogical tree at site $i$,
the sum is over edges $e$ in the tree, $\ell(e)$ is the length of the edge $e$,
and $x_j(e)$ is the number of individuals in $A_j$ that fall below $e$.

\emph{Note:} somewhat more generally,
we could assign a set of weights to the individuals in each group,
and let $x_j$ denote the total weight of the individuals in $A_j$ 
inheriting from that allele or edge.
For simplicity we do not pursue this here.
\plr{this is a good use of the word ``weight'' (although these could also be negative)}


\paragraph{Computation}
The \emph{tree differences} encoded by a tree sequence allow us to efficiently update the quantities $x$ used above.
As it is more straightforward, we first describe the algorithm
to compute the branch length version.

\plr{Rework these into three algorithms: one to maintain the node weights $x$ 
    (basically already described as `tracked nodes' in msprime)
    and the other two that use this.}

\paragraph{Total branch length weight algorithm}
In a tree sequence, every branch in any tree is associated with its terminal \emph{node}.
We will keep updated an array $x$ so that for any node $e$,
the value $x[e][j]$ is the number of samples in $A_j$ below node $e$ in the current tree.
This will be zero for any node \emph{not} appearing in the current tree.
We also maintain $T$, the total weight of the current tree.
\begin{itemize}
    \item \emph{Initialize} all elements of $x$ to zero
        except that $x[a][j] = 1$ for each $a \in A_j$ and $1 \le j \le k$.
        Also set $S=0$ and $T=0$.
    \item Pop the next \emph{tree difference} off the stack: an interval with endpoints $\ell < r$,
        a set of edges $\mathcal{R}$ to remove, and a set of edges $\mathcal{A}$ to add.
    \item \emph{Remove} each edge (parent, child) in $\mathcal{R}$ from the tree,
        let $u$ = parent, and while $u != -1$,
        subtract $x[child]$ from $x[u]$ and set $u$ to be parent($u$).
    \item \emph{Add} each edge (parent, child) in $\mathcal{R}$ to the tree,
        let $u$ = parent, and while $u != -1$,
        add $x[child]$ to $x[u]$ and set $u$ to be parent($u$).
    \item For each node that has changed, let $w_\text{old}$ and $w_\text{new}$
        be the weights before and after the change in $x$,
        and add $w_\text{new} - w_text{old}$ to $T$.
    \item Add $T \times (r-\ell)$ to the running total, $S$.
    \item Return to (2).
\end{itemize}
\plr{should modify code to maintain a vector of $w$ values}
The average weighted branch length is obtained by dividing by the total number of sites.

The algorithm for sites is similar, also maintaining the array $x$,
and making use of the association of each mutation with a particular node in the tree.

\paragraph{Total site weight algorithm}
This works as the \emph{total branch length weight algorithm} except that steps (5--6) are replaced with:
\plr{gah this is terrible}
\begin{itemize}
    \item[5'] \emph{Find allele weights} for each mutated site in the tree:
        if $a$ is the ancestral allele,
        set $u[a]$ to be the value of $x$ at the root; and then,
        for each mutation at node $e$ with derived allele $b$ 
        and parent mutation occurring at node $p$ with derived allele $c$,
        add $x[e]$ to $u[b]$, and subtract $x[p]$ from $u[c]$.
    \item[6'] For each unique allele $a$ seen,
        add $w(v[a])$ to $T$.
\end{itemize}
\plr{check that algorithm deals correctly with more than one mut per node}

%%%%%%%%%%%%%%%%%%
\subsection*{General properties}

Here we describe some properties that we would like our statistics to have.

\paragraph{They do not depend on choice of reference allele.}
This is achieved by applying the \emph{same} weighting function to every allele.
Ancestral information can be incorporated by including an outgroup or ancestral sequence
as an additional group of size one.
Equivalently, we always work with \emph{unrooted} trees.

\paragraph{They do not depend on sample size.}
We'd like our statistics to be \emph{consistent estimators}
of a single quantity, and should therefore not depend on sample size in expectation.
\plr{insert this above}
This is achieved by always computing probabilies of sampling \emph{without replacement},
i.e., replacing each occurence of $p^k$ with $(x)_k/(n)_k$.
This lets us compare values of a given statistic
that are computed on different data sets with different sample sizes.

\paragraph{Equivalence of ``site'' and ``branch length'' statistics}
The definitions above are made so that for any weighting function,
$\site_w$ is a moment estimator of $\branch_w$, i.e.,
that $\E[\site_w] = \branch_w$ for \plr{somehow say continuum of sites here}.


%%%%%%%%%%%%%%%%%%%%%%%
\section*{Results}



%%%%%%%%%%%%%%%%%%%%%%%
\section*{Discussion}



%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Acknowledgement}


\bibliography{references}

\end{document}
